### PR TITLE
More wgsl-out improvements

### PR DESCRIPTION
**Pull request by Gordon-F**
1. Return Err instead of panic for unimplemented cases
2. Implement writing for Scalar, Array, Matrix
3. Share some code between backends

### CODE REVIEW

File: src/back/wgsl/writer.rs [hunk 1]
```diff
@@ -463,30 +517,7 @@ impl<W: Write> Writer<W> {
                 self.write_expr(module, left, func_ctx)?;
 
                 // TODO: glsl-out copy-paste
```
Comment (ColdIce1605) [inline src/back/wgsl/writer.rs:519]: I think this todo can be removed.

File: src/back/glsl/mod.rs [hunk 2]
```diff
@@ -1836,30 +1836,7 @@ impl<'a, W: Write> Writer<'a, W> {
...
-                            BinaryOperator::And => "&",
-                            BinaryOperator::ExclusiveOr => "^",
-                            BinaryOperator::InclusiveOr => "|",
-                            BinaryOperator::LogicalAnd => "&&",
-                            BinaryOperator::LogicalOr => "||",
-                            BinaryOperator::ShiftLeft => "<<",
-                            BinaryOperator::ShiftRight => ">>",
-                        }
-                    )?;
+                    write!(self.out, " {} ", crate::back::binary_operation_str(op))?;
```
Comment (kvark) [inline src/back/glsl/mod.rs:1839]: nit: maybe refer to this as `super::` instead?

File: src/back/mod.rs [hunk 3]
```diff
@@ -33,3 +33,53 @@ impl crate::Expression {
         }
     }
 }
+
+/// Helper function that returns the string corresponding to the [`BinaryOperator`](crate::BinaryOperator)
+/// # Notes
+/// Used by `glsl-out`, `msl-out`, `wgsl-out`.
+fn binary_operation_str(op: crate::BinaryOperator) -> &'static str {
+    match op {
+        crate::BinaryOperator::Add => "+",
```
Comment (kvark) [inline src/back/mod.rs:42]: let's put `use crate::BinaryOperator as Bo` at the top of the function

File: src/back/mod.rs [hunk 4]
```diff
@@ -33,3 +33,53 @@ impl crate::Expression {
...
+        crate::VectorSize::Bi => "2",
+        crate::VectorSize::Tri => "3",
+        crate::VectorSize::Quad => "4",
+    }
+}
+
+/// Helper function that returns the string corresponding to the [`ImageDimension`](crate::ImageDimension)
+/// # Notes
+/// Used by `msl-out`, `wgsl-out`.
+fn image_dimension_str(dim: crate::ImageDimension) -> &'static str {
```
Comment (kvark) [inline src/back/mod.rs:78]: let's not have this shared, it seems too language specific

**Review by kvark — COMMENTED** [review]
Nice work!

**Comment by Gordon-F** [comment]
@kvark Is there some way to disable clippy warning only for some features? To avoid warning like this one: `function is never used: binary_operation_str`

**Comment by kvark** [comment]
@Gordon-F see https://github.com/gfx-rs/naga/blob/77a64da189e6ed2c4631ca03cb01b8d8124c4dad/src/front/mod.rs#L22

**Comment by Gordon-F** [comment]
> @Gordon-F see
> 
> https://github.com/gfx-rs/naga/blob/77a64da189e6ed2c4631ca03cb01b8d8124c4dad/src/front/mod.rs#L22

Thank you! PR is ready to merge.

**Review by kvark — APPROVED** [review]
Thank you!

**Pull request merged**